tests/AdapterTest.php

```diff
@@ -1,50 +1,69 @@
-        $result = method_exists($stmt, 'fetchAssociative') ? $stmt->fetchAllAssociative() : $stmt->fetchAll();
-
-        $result = array_map([$adapter, "filterRule"], $result);
-
-        $this->assertEquals($newPolicies, $result);
-    }
-}
+<?php
+
+namespace CasbinAdapter\DBAL\Tests;
+
+use CasbinAdapter\DBAL\Adapter as DatabaseAdapter;
+use Doctrine\DBAL\Configuration;
+use Doctrine\DBAL\DriverManager;
+use Doctrine\DBAL\Exception;
+use Doctrine\DBAL\Logging\Middleware as LoggingMiddleware;
+
+class AdapterTest extends TestCase
+{
+    /**
+     * @throws Exception
+     */
+    public function testUpdateFilteredPolicies()
+    {
+        $this->initConfig();
+        $connConfig = new Configuration();
+        $this->configureLogger($connConfig);
+        $conn = DriverManager::getConnection(
+            $this->config,
+            $connConfig
+        );
+        $adapter = DatabaseAdapter::newAdapter($conn);
+        $conn->delete($adapter->policyTableName, ["1" => "1"]);
+        $conn->insert($adapter->policyTableName, ["p_type" => "p", "v0" => "alice", "v1" => "data", "v2" => "read", "v3" => "allow"]);
+        $conn->insert($adapter->policyTableName, ["p_type" => "p", "v0" => "alice", "v1" => "data1", "v2" => "write", "v3" => "allow"]);
+        $conn->insert($adapter->policyTableName, ["p_type" => "p", "v0" => "bob", "v1" => "data", "v2" => "write", "v3" => "allow"]);
+        $conn->insert($adapter->policyTableName, ["p_type" => "p", "v0" => "bob", "v1" => "data1", "v2" => "read", "v3" => "allow"]);
+
+        $newPolicies = [
+            ["alice", "data", "read", "allow"],
+            ["bob", "data", "read", "allow"]
+        ];
+        $oldRules = $adapter->updateFilteredPolicies("p", "p", $newPolicies, 1, "data", null, "allow");
+        $this->assertEquals([
+            ["p", "alice", "data", "read", "allow"],
+            ["p", "bob", "data", "write", "allow"]
+        ], $oldRules);
+
+        $query = $conn->createQueryBuilder()->from($adapter->policyTableName)->where('p_type = "p" and v1 = "data" and v3 = "allow"')->select("v0", "v1", "v2", "v3");
+        $stmt = method_exists($query, "executeQuery") ? $query->executeQuery() : $query->execute();
+        $result = method_exists($stmt, 'fetchAssociative') ? $stmt->fetchAllAssociative() : $stmt->fetchAll();
+
+        $result = array_map([$adapter, "filterRule"], $result);
+
+        $this->assertEquals($newPolicies, $result);
+    }
+
+    /**
+     *
+     * @param \Doctrine\DBAL\Configuration $connConfig
+     * @return void
+     */
+    private function configureLogger($connConfig)
+    {
+        // Doctrine < 4.0
+        if(method_exists($connConfig, "setSQLLogger")) {
+            $connConfig->setSQLLogger(new DebugStackLogger());
+        }
+        // Doctrine >= 4.0
+        else {
+            $connConfig->setMiddlewares([
+              new LoggingMiddleware(new PsrLogger())
+            ]);
+        }
+    }
+}
```

Comment:
Please modify, do not change all lines

Reply:
This is due to different line endings (`CRLF` vs `LF`). I can try to fix it if you let me know which line endings this repository prefers, since there's no `.gitattributes` file.

The weird thing is that `Adapter.php` didn't show all lines as changed. Does this mean that test have a different line ending than the source code? I highly recommend adding a `.gitattributes` and normalizing all line endings if that's the case.

Reply:
For the sake of reviewing the actual changes, I've attached diffs without whitespace changes for both files below:

```diff
diff --git a/tests/AdapterTest.php b/tests/AdapterTest.php
index b27f34a..1b5eb2a 100644
--- a/tests/AdapterTest.php
+++ b/tests/AdapterTest.php
@@ -6,10 +6,10 @@ use CasbinAdapter\DBAL\Adapter as DatabaseAdapter;
 use Doctrine\DBAL\Configuration;
 use Doctrine\DBAL\DriverManager;
 use Doctrine\DBAL\Exception;
+use Doctrine\DBAL\Logging\Middleware as LoggingMiddleware;
 
 class AdapterTest extends TestCase
 {
-
     /**
      * @throws Exception
      */
@@ -17,8 +17,7 @@ class AdapterTest extends TestCase
     {
         $this->initConfig();
         $connConfig = new Configuration();
-        $logger = new DebugStackLogger();
-        $connConfig->setSQLLogger($logger);
+        $this->configureLogger($connConfig);
         $conn = DriverManager::getConnection(
             $this->config,
             $connConfig
@@ -40,11 +39,31 @@ class AdapterTest extends TestCase
             ["p", "bob", "data", "write", "allow"]
         ], $oldRules);
 
-        $stmt = $conn->createQueryBuilder()->from($adapter->policyTableName)->where('p_type = "p" and v1 = "data" and v3 = "allow"')->select("v0", "v1", "v2", "v3")->execute();
+        $query = $conn->createQueryBuilder()->from($adapter->policyTableName)->where('p_type = "p" and v1 = "data" and v3 = "allow"')->select("v0", "v1", "v2", "v3");
+        $stmt = method_exists($query, "executeQuery") ? $query->executeQuery() : $query->execute();
         $result = method_exists($stmt, 'fetchAssociative') ? $stmt->fetchAllAssociative() : $stmt->fetchAll();
 
         $result = array_map([$adapter, "filterRule"], $result);
 
         $this->assertEquals($newPolicies, $result);
     }
+
+    /**
+     *
+     * @param \Doctrine\DBAL\Configuration $connConfig
+     * @return void
+     */
+    private function configureLogger($connConfig)
+    {
+        // Doctrine < 4.0
+        if(method_exists($connConfig, "setSQLLogger")) {
+            $connConfig->setSQLLogger(new DebugStackLogger());
+        }
+        // Doctrine >= 4.0
+        else {
+            $connConfig->setMiddlewares([
+              new LoggingMiddleware(new PsrLogger())
+            ]);
+        }
+    }
 }
```

```diff
diff --git a/tests/TestCase.php b/tests/TestCase.php
index f5371fa..eeb21b7 100644
--- a/tests/TestCase.php
+++ b/tests/TestCase.php
@@ -32,7 +32,8 @@ class TestCase extends \PHPUnit\Framework\TestCase
         $tableName = $adapter->policyTableName;
         $conn = $adapter->getConnection();
         $queryBuilder = $conn->createQueryBuilder();
-        $queryBuilder->delete($tableName)->where('1 = 1')->execute();
+        $query = $queryBuilder->delete($tableName)->where('1 = 1');
+        method_exists($query, "executeQuery") ? $query->executeQuery() : $query->execute();
 
         $data = [
             ['p_type' => 'p', 'v0' => 'alice', 'v1' => 'data1', 'v2' => 'read'],
@@ -42,7 +43,8 @@ class TestCase extends \PHPUnit\Framework\TestCase
             ['p_type' => 'g', 'v0' => 'alice', 'v1' => 'data2_admin'],
         ];
         foreach ($data as $row) {
-            $queryBuilder->insert($tableName)->values(array_combine(array_keys($row), array_fill(0, count($row), '?')))->setParameters(array_values($row))->execute();
+            $query = $queryBuilder->insert($tableName)->values(array_combine(array_keys($row), array_fill(0, count($row), '?')))->setParameters(array_values($row));
+            method_exists($query, "executeQuery") ? $query->executeQuery() : $query->execute();
         }
     }
```